key error bug when invoking pex without arguments

```
$ ./pants -ldebug pex
15:55:18.96 [DEBUG] acquiring lock: <pants.pantsd.lock.OwnerPrintingInterProcessFileLock object at 0x10a7f1f70>
15:55:18.96 [DEBUG] releasing lock: <pants.pantsd.lock.OwnerPrintingInterProcessFileLock object at 0x10a7f1f70>
15:55:18.96 [DEBUG] Connecting to pantsd on port 62729
15:55:18.97 [DEBUG] Connecting to pantsd on port 62729 attempt 1/3
15:55:18.97 [DEBUG] Connected to pantsd
15:55:19.00 [DEBUG] Launching 1 roots (poll=false).
15:55:19.02 [DEBUG] Starting: pants.backend.python.util_rules.pex.build_pex
15:55:19.02 [DEBUG] Starting: Find Python interpreter to bootstrap PEX
15:55:19.02 [DEBUG] Canceled: pants.backend.python.util_rules.pex.build_pex
15:55:19.02 [DEBUG] Starting: Find Python interpreter for constraints
15:55:19.02 [DEBUG] Canceled: Find Python interpreter for constraints
15:55:19.02 [DEBUG] Starting: Find Python interpreter to bootstrap PEX
15:55:19.02 [DEBUG] Completed: Find Python interpreter to bootstrap PEX - Selected /Users/aadt/.cache/pants/pants_dev_deps/Darwin.x86_64.i386.py38.venv/bin/python to bootstrap PEXes with.
15:55:19.02 [DEBUG] Completed: Find Python interpreter to bootstrap PEX - Selected /Users/aadt/.cache/pants/pants_dev_deps/Darwin.x86_64.i386.py38.venv/bin/python to bootstrap PEXes with.
15:55:19.02 [DEBUG] Starting: pants.backend.python.util_rules.pex.build_pex
15:55:19.02 [DEBUG] Starting: Find Python interpreter for constraints
15:55:19.02 [DEBUG] Completed: Find Python interpreter for constraints - Selected /usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/bin/python3.8 to run PEXes with.
15:55:19.02 [DEBUG] Completed: pants.backend.python.util_rules.pex.build_pex
15:55:19.02 [DEBUG] computed 1 nodes in 0.028035 seconds. there are 127 total nodes.
15:55:19.07 [DEBUG] Launching 1 roots (poll=false).
15:55:19.07 [DEBUG] computed 1 nodes in 0.000628 seconds. there are 127 total nodes.
15:55:19.13 [DEBUG] specs are: Specs(address_specs=AddressSpecs(literals=(), globs=(), filter_by_global_options=True), filesystem_specs=FilesystemSpecs(includes=(), ignores=()))
15:55:19.13 [DEBUG] changed_options are: ChangedOptions(since=None, diffspec=None, dependees=<DependeesOption.NONE: 'none'>)
15:55:19.13 [DEBUG] Launching 1 roots (poll=false).
15:55:19.13 [DEBUG] computed 1 nodes in 0.004937 seconds. there are 788 total nodes.
15:55:19.13 [DEBUG] BuildSense plugin is disabled.
15:55:19.13 [ERROR] Exception caught: (builtins.KeyError)
  File "/Users/aadt/src/github/kaos/pants/src/python/pants/bin/local_pants_runner.py", line 234, in _run_inner
    return self._perform_run(goals)
  File "/Users/aadt/src/github/kaos/pants/src/python/pants/bin/local_pants_runner.py", line 173, in _perform_run
    return self._perform_run_body(goals, poll=False)
  File "/Users/aadt/src/github/kaos/pants/src/python/pants/bin/local_pants_runner.py", line 190, in _perform_run_body
    return self.graph_session.run_goal_rules(
  File "/Users/aadt/src/github/kaos/pants/src/python/pants/init/engine_initializer.py", line 122, in run_goal_rules
    goal_product = self.goal_map[goal]

Exception message: 'pex'
Traceback (most recent call last):
  File "/Users/aadt/src/github/kaos/pants/src/python/pants/bin/local_pants_runner.py", line 234, in _run_inner
    return self._perform_run(goals)
  File "/Users/aadt/src/github/kaos/pants/src/python/pants/bin/local_pants_runner.py", line 173, in _perform_run
    return self._perform_run_body(goals, poll=False)
  File "/Users/aadt/src/github/kaos/pants/src/python/pants/bin/local_pants_runner.py", line 190, in _perform_run_body
    return self.graph_session.run_goal_rules(
  File "/Users/aadt/src/github/kaos/pants/src/python/pants/init/engine_initializer.py", line 122, in run_goal_rules
    goal_product = self.goal_map[goal]
KeyError: 'pex'

(Use --no-process-execution-local-cleanup to inspect chroots. See https://www.pantsbuild.org/v2.7/docs/troubleshooting for common issues. Consider reaching out for help: https://www.pantsbuild.org/v2.7/docs/getting-help.)
```

--- a/src/python/pants/engine/goal.py
+++ b/src/python/pants/engine/goal.py
@@ -8,6 +8,7 @@ from typing import TYPE_CHECKING, Callable, ClassVar, Iterator, Tuple, Type, cas
 
 from typing_extensions import final
 
+from pants.option.scope import ScopeInfo
 from pants.option.subsystem import Subsystem
 from pants.util.meta import classproperty
 
@@ -36,6 +37,10 @@ class GoalSubsystem(Subsystem):
     # If the goal requires downstream implementations to work properly, such as `test` and `run`,
     # it should declare the union types that must have members.
     required_union_implementations: Tuple[Type, ...] = ()
+
+    @classmethod
+    def create_scope_info(cls, **scope_info_kwargs) -> ScopeInfo:
+        return super().create_scope_info(is_goal=True, **scope_info_kwargs)
 
     @classproperty
     @abstractmethod

--- a/src/python/pants/engine/goal_test.py
+++ b/src/python/pants/engine/goal_test.py
@@ -4,6 +4,7 @@
 from pants.engine.console import Console
 from pants.engine.goal import Goal, GoalSubsystem, LineOriented
 from pants.engine.rules import goal_rule
+from pants.option.scope import ScopeInfo
 from pants.testutil.option_util import create_goal_subsystem, create_options_bootstrapper
 from pants.testutil.rule_runner import mock_console, run_rule_with_mocks
 
@@ -33,3 +34,14 @@ def test_line_oriented_goal() -> None:
         )
         assert result.exit_code == 0
         assert stdio_reader.get_stdout() == "output...line oriented\n"
+
+
+def test_goal_scope_flag() -> None:
+    class DummyGoal(GoalSubsystem):
+        name = "dummy"
+
+    dummy = create_goal_subsystem(DummyGoal)
+    assert dummy.get_scope_info().is_goal is True
+    assert dummy.known_scope_infos() == {
+        ScopeInfo(scope="dummy", optionable_cls=DummyGoal, is_goal=True)
+    }

--- a/src/python/pants/option/arg_splitter.py
+++ b/src/python/pants/option/arg_splitter.py
@@ -84,7 +84,7 @@ class ArgSplitter:
     def __init__(self, known_scope_infos: Iterable[ScopeInfo], buildroot: str) -> None:
         self._buildroot = Path(buildroot)
         self._known_scope_infos = known_scope_infos
-        self._known_scopes = {si.scope for si in known_scope_infos} | {
+        self._known_scopes = {si.scope for si in known_scope_infos if si.is_goal} | {
             "version",
             "help",
             "help-advanced",

--- a/src/python/pants/option/arg_splitter.py
+++ b/src/python/pants/option/arg_splitter.py
@@ -84,12 +84,13 @@ class ArgSplitter:
     def __init__(self, known_scope_infos: Iterable[ScopeInfo], buildroot: str) -> None:
         self._buildroot = Path(buildroot)
         self._known_scope_infos = known_scope_infos
-        self._known_scopes = {si.scope for si in known_scope_infos if si.is_goal} | {
+        self._known_scopes = {si.scope for si in known_scope_infos} | {
             "version",
             "help",
             "help-advanced",
             "help-all",
         }
+        self._known_goal_scopes = {si.scope for si in known_scope_infos if si.is_goal}
         self._unconsumed_args: List[
             str
         ] = []  # In reverse order, for efficient popping off the end.
@@ -171,7 +172,8 @@ class ArgSplitter:
         while scope:
             if not self._check_for_help_request(scope.lower()):
                 add_scope(scope)
-                goals.add(scope.partition(".")[0])
+                if scope in self._known_goal_scopes:
+                    goals.add(scope.partition(".")[0])
                 for flag in flags:
                     assign_flag_to_scope(flag, scope)
             scope, flags = self._consume_scope()

--- a/src/python/pants/option/arg_splitter.py
+++ b/src/python/pants/option/arg_splitter.py
@@ -174,6 +174,8 @@ class ArgSplitter:
                 add_scope(scope)
                 if scope in self._known_goal_scopes:
                     goals.add(scope.partition(".")[0])
+                else:
+                    unknown_scopes.append(scope)
                 for flag in flags:
                     assign_flag_to_scope(flag, scope)
             scope, flags = self._consume_scope()

--- a/src/python/pants/option/arg_splitter_test.py
+++ b/src/python/pants/option/arg_splitter_test.py
@@ -22,15 +22,18 @@ from pants.util.contextutil import pushd, temporary_dir
 
 class ArgSplitterTest(unittest.TestCase):
     _known_scope_infos = [
-        ScopeInfo("compile"),
-        ScopeInfo("compile.java"),
-        ScopeInfo("compile.scala"),
-        ScopeInfo("jvm"),
-        ScopeInfo("jvm.test.junit"),
-        ScopeInfo("reporting"),
-        ScopeInfo("test"),
-        ScopeInfo("test.junit"),
-    ]
+        ScopeInfo(scope, is_goal=True)
+        for scope in [
+            "compile",
+            "compile.java",
+            "compile.scala",
+            "jvm",
+            "jvm.test.junit",
+            "reporting",
+            "test",
+            "test.junit",
+        ]
+    ] + [ScopeInfo("hidden", is_goal=False)]
 
     @staticmethod
     def assert_valid_split(
@@ -384,3 +387,6 @@ class ArgSplitterTest(unittest.TestCase):
         splitter = ArgSplitter(ArgSplitterTest._known_scope_infos, buildroot=os.getcwd())
         splitter.split_args(shlex.split("./pants foo/bar:baz"))
         self.assertTrue(isinstance(splitter.help_request, NoGoalHelp))
+
+    def test_hidden_scope_is_unknown_goal(self) -> None:
+        self.assert_unknown_goal("./pants hidden", ["hidden"])

--- a/src/python/pants/option/optionable.py
+++ b/src/python/pants/option/optionable.py
@@ -95,11 +95,16 @@ class Optionable(OptionableFactory, metaclass=ABCMeta):
             )
 
     @classmethod
+    def create_scope_info(cls, **scope_info_kwargs) -> ScopeInfo:
+        """One place to create scope info, to allow subclasses to inject custom scope args."""
+        return ScopeInfo(**scope_info_kwargs)
+
+    @classmethod
     def get_scope_info(cls) -> ScopeInfo:
         """Returns a ScopeInfo instance representing this Optionable's options scope."""
         if cls.options_scope is None:
             raise OptionsError(f"{cls.__name__} must set options_scope.")
-        return ScopeInfo(scope=cast(str, cls.options_scope), optionable_cls=cls)
+        return cls.create_scope_info(scope=cast(str, cls.options_scope), optionable_cls=cls)
 
     @classmethod
     def subscope(cls, scope) -> str:

--- a/src/python/pants/option/options_test.py
+++ b/src/python/pants/option/options_test.py
@@ -57,7 +57,7 @@ def global_scope() -> ScopeInfo:
 
 
 def task(scope: str) -> ScopeInfo:
-    return ScopeInfo(scope)
+    return ScopeInfo(scope, is_goal=True)
 
 
 def intermediate(scope: str) -> ScopeInfo:

--- a/src/python/pants/option/scope.py
+++ b/src/python/pants/option/scope.py
@@ -34,6 +34,9 @@ class ScopeInfo:
     removal_version: Optional[str] = None
     removal_hint: Optional[str] = None
 
+    # command line goal scope flag
+    is_goal: bool = False
+
     @property
     def description(self) -> str:
         return cast(str, getattr(self.optionable_cls, "help"))

--- a/src/python/pants/option/subsystem.py
+++ b/src/python/pants/option/subsystem.py
@@ -96,7 +96,7 @@ class Subsystem(Optionable):
         if subscope is None:
             return super().get_scope_info()
         else:
-            return ScopeInfo(cls.subscope(subscope), cls)
+            return cls.create_scope_info(scope=cls.subscope(subscope), optionable_cls=cls)
 
     def __init__(self, scope: str, options: OptionValueContainer) -> None:
         super().__init__()
@@ -191,9 +191,9 @@ class Subsystem(Optionable):
                 if scoped_to == GLOBAL_SCOPE
                 else optionable_cls.subscope(scoped_to)
             )
-            scope_info = ScopeInfo(
-                scope,
-                optionable_cls,
+            scope_info = cls.create_scope_info(
+                scope=scope,
+                optionable_cls=optionable_cls,
                 removal_version=removal_version,
                 removal_hint=removal_hint,
             )

